node/core/approval-voting/src/ops.rs

```diff
@@ -311,3 +311,74 @@ pub fn force_approve(
 
 	Ok(approved_hashes)
 }
+
+/// Revert to the block corresponding to the specified `hash`.
+/// The operation is not allowed for blocks older than the last finalized one.
+pub fn revert_to(
+	overlay: &mut OverlayedBackend<'_, impl Backend>,
+	hash: Hash,
+) -> SubsystemResult<()> {
+	let (children, height) = match overlay.load_block_entry(&hash)? {
+		Some(mut entry) => {
+			let children_height = entry.block_number() + 1;
+			let children = std::mem::take(&mut entry.children);
+			// Write revert point block entry without the children.
+			overlay.write_block_entry(entry);
+			(children, children_height)
+		},
+		None => {
+			// May be a revert to the last finalized block.
+			// Loading all blocks just to get the first one may be expensive
+			// but we should consider that revert is a "one-shot" and very
+			// sporadic operation.
+			let blocks = overlay.load_all_blocks()?;
+			let child_entry = blocks
+				.first()
+				.and_then(|hash| overlay.load_block_entry(hash).ok())
+				.flatten()
+				.ok_or_else(|| {
+					SubsystemError::Context(format!("lookup failure for block {}", hash))
+				})?;
+
+			// The parent is expected to be the revert point
+			if child_entry.parent_hash() != hash {
+				return Err(SubsystemError::Context(
+					"revert below last finalized block or corrupted storage".to_string(),
+				))
+			}
+
+			let children_height = child_entry.block_number();
```

Comment:
why not use `load_stored_blocks()` to get the height of the first stored block?

Reply:
also, there might be multiple blocks stored at the bottom height

Reply:
Good catch. Thank you